pandas/core/apply.py

```diff
@@ -308,11 +312,19 @@ def apply_standard(self):
                 # reached via numexpr; fall back to python implementation
                 pass
             else:
-                return self.obj._constructor_sliced(result, index=labels)
+                return_result = self.obj._constructor_sliced(result, index=labels)
+                if self.axis != 0 and self.axis != "index":
+                    return return_result
 
         # compute the result using the series generator
         results, res_index = self.apply_series_generator()
 
+        if flag and return_result is not None:
```

Comment:
what is this case tryng to catch?

Reply:
There are 2 cases
1) can_reduce is False - the function doesn't change its behaviour, compared to previous version
2) can_reduce is True - the return_result is calculated like before, but it's returned only if we don't use apply on columns or we don't have mixed dtypes. In other case I use apply_series_generator to obtain column-by-column result and after that I'm wrapping it using _constructor_sliced. It's done this way to preserve the "old" way - it was solution to problem with my function sometimes returning Series instead of DataFrame (or the other way). I assumed, that already existing way of getting result would be better, than thinking about some new solutions.

Reply:
@jreback